trie/leavesRetriever/leavesRetriever.go

```diff
@@ -0,0 +1,153 @@
+package leavesRetriever
+
+import (
+	"context"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/hashing"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"github.com/multiversx/mx-chain-go/common"
+	"github.com/multiversx/mx-chain-go/trie/leavesRetriever/dfsTrieIterator"
+)
+
+type leavesRetriever struct {
+	iterators      map[string]common.DfsIterator
+	lruIteratorIDs [][]byte
+	db             common.TrieStorageInteractor
+	marshaller     marshal.Marshalizer
+	hasher         hashing.Hasher
+	size           uint64
+	maxSize        uint64
+	mutex          sync.RWMutex
+}
+
+// NewLeavesRetriever creates a new leaves retriever
+func NewLeavesRetriever(db common.TrieStorageInteractor, marshaller marshal.Marshalizer, hasher hashing.Hasher, maxSize uint64) (*leavesRetriever, error) {
+	if check.IfNil(db) {
+		return nil, ErrNilDB
+	}
+	if check.IfNil(marshaller) {
+		return nil, ErrNilMarshaller
+
+	}
+	if check.IfNil(hasher) {
+		return nil, ErrNilHasher
+	}
+
+	return &leavesRetriever{
+		iterators:      make(map[string]common.DfsIterator),
+		lruIteratorIDs: make([][]byte, 0),
+		db:             db,
+		marshaller:     marshaller,
+		hasher:         hasher,
+		size:           0,
+		maxSize:        maxSize,
+	}, nil
+}
+
+// GetLeaves retrieves the leaves from the trie. If there is a saved checkpoint for the iterator id, it will continue to iterate from the checkpoint.
+func (lr *leavesRetriever) GetLeaves(numLeaves int, rootHash []byte, iteratorID []byte, ctx context.Context) (map[string]string, []byte, error) {
+	if len(iteratorID) == 0 {
+		return lr.getLeavesFromNewInstance(numLeaves, rootHash, ctx)
+	}
+
+	lr.mutex.RLock()
+	iterator, ok := lr.iterators[string(iteratorID)]
+	lr.mutex.RUnlock()
+	if !ok {
+		return nil, nil, ErrIteratorNotFound
+	}
+
+	return lr.getLeavesFromCheckpoint(numLeaves, iterator, iteratorID, ctx)
+}
+
+func (lr *leavesRetriever) getLeavesFromNewInstance(numLeaves int, rootHash []byte, ctx context.Context) (map[string]string, []byte, error) {
+	iterator, err := dfsTrieIterator.NewIterator(rootHash, lr.db, lr.marshaller, lr.hasher)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return lr.getLeavesFromIterator(iterator, numLeaves, ctx)
+}
+
+func (lr *leavesRetriever) getLeavesFromCheckpoint(numLeaves int, iterator common.DfsIterator, iteratorID []byte, ctx context.Context) (map[string]string, []byte, error) {
```

Comment:
They are related, but every time you get the `iteratorID` from `iterator` there are some computations that take place.  The `iteratorID` is computed based on the contents of the iterator. Would leave like this in order to avoid doing the same computation multiple times.